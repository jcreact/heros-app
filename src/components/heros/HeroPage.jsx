import React, { useMemo } from 'react';
import { Redirect, useParams } from 'react-router-dom';
import { createMuiTheme, ThemeProvider } from '@material-ui/core/styles';
import { red, blue } from '@material-ui/core/colors';
import {
    Avatar,
    Box,
    Button,
    Card,
    CardMedia,
    Chip,
    Divider,
    Fade,
    Grid,
    makeStyles,
    Typography,
} from '@material-ui/core';
import { ArrowBackIos } from '@material-ui/icons';

import { getHeroById } from '../../selectors/getHeroById';

const useStyles = makeStyles((theme) => ({
    chips: {
        margin: theme.spacing(0.5),
    },
}));

export const HeroPage = ({ history }) => {
    const { chips } = useStyles();
    const { id } = useParams();
    const hero = getHeroById(id);

    const theme = useMemo(() => {
        if (!hero) {
            return createMuiTheme();
        }
        if (hero?.publisher.startsWith('DC')) {
            return createMuiTheme({
                palette: {
                    primary: blue,
                },
            });
        } else {
            return createMuiTheme({
                palette: {
                    primary: red,
                },
            });
        }
    }, [hero]);

    if (!hero) {
        return <Redirect to="/" />;
    }

    const { superhero, publisher, alter_ego, first_appearance, characters } =
        hero;

    const charactersChips = () => {
        return characters
            .split(',')
            .map((c) => [
                c
                    .match(/\b(\w)/g)
                    .join('')
                    .toUpperCase(),
                c,
            ])
            .map(([x, c], i) => (
                <Chip
                    key={i}
                    label={c}
                    color="default"
                    variant="outlined"
                    className={chips}
                    avatar={<Avatar>{x}</Avatar>}
                />
            ));
    };

    const handleClick = (_) => {
        if (history.length <= 2) {
            const back = `/${publisher.startsWith('DC') ? 'dc' : 'marvel'}`;
            console.log(back);
            history.push(back);
        } else {
            history.goBack();
        }
    };

    return (
        <Fade in timeout={400}>
            <Box
                display="flex"
                justifyContent="center"
                alignContent="center"
                mt={4}
            >
                <Grid container spacing={2} style={{ maxWidth: '110vh' }}>
                    <Grid item xs={12} sm={6}>
                        <Card variant="outlined">
                            <CardMedia
                                image={`./assets/heros/${id}.jpg`}
                                component="img"
                            />
                        </Card>
                    </Grid>
                    <Grid item xs={12} sm={6}>
                        <Box
                            display="flex"
                            flexDirection="column"
                            justifyContent="flex-start"
                            style={{ height: '100%' }}
                        >
                            <Typography variant="h2">
                                <span>{superhero}</span>
                                <ThemeProvider theme={theme}>
                                    <Box clone ml={1}>
                                        <Chip
                                            size="small"
                                            label={publisher}
                                            variant="outlined"
                                            color="primary"
                                        />
                                    </Box>
                                </ThemeProvider>
                            </Typography>
                            <Box mt={2} mb={2}>
                                <Divider />
                            </Box>
                            <Box mt={2}>
                                <Typography variant="h5">
                                    <b style={{ textTransform: 'capitalize' }}>
                                        Alter ego:{' '}
                                    </b>
                                    {alter_ego}
                                </Typography>
                            </Box>
                            <Box mt={2} css={{ flexGrow: 1 }}>
                                <Typography variant="h5">
                                    <b style={{ textTransform: 'capitalize' }}>
                                        First Appearance:{' '}
                                    </b>
                                    {first_appearance}
                                </Typography>
                            </Box>
                            <Box
                                display="flex"
                                justifyContent="center"
                                flexWrap="wrap"
                                mt={2}
                            >
                                {charactersChips()}
                            </Box>
                            <Box mt={4} mb={2}>
                                <Divider />
                            </Box>
                            <Box
                                mt={2}
                                display="flex"
                                justifyContent="flex-end"
                            >
                                <Button
                                    color="primary"
                                    startIcon={<ArrowBackIos />}
                                    onClick={handleClick}
                                >
                                    regresar
                                </Button>
                            </Box>
                        </Box>
                    </Grid>
                </Grid>
            </Box>
        </Fade>
    );
};
