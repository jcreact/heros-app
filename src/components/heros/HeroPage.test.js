import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';

// Com
import { HeroPage } from './HeroPage';


describe(`Pruebas en el componente 'HeroPage.jsx'.`, () => {

    const history = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn()
    };

    test(`Debe mostrar el componente 'Redirect' si no hay argumentos en el URL.`, () => {
        const component = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroPage history={history} />
            </MemoryRouter>
        );

        expect(component.find('Redirect').exists()).toBeTruthy();
    });

    test(`Debe mostrar a Superman.`, () => {
        const component = mount(
            <MemoryRouter initialEntries={['/hero/dc-superman']}>
                <Route path="/hero/:id" component={HeroPage} />
            </MemoryRouter>
        );

        expect(component).toMatchSnapshot();
        expect(component.find('span').at(0).text()).toBe('Superman');
        expect(component.find('span').at(1).text()).toBe('DC Comics');
        expect(component.find('h5.MuiTypography-h5').at(0).text()).toContain('Kal-El');
        expect(component.find('h5.MuiTypography-h5').at(1).text()).toContain('Action Comics #1');
    });

    test(`Debe regresar a la página anterior '/dc' con push().`, () => {
        const history = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn()
        };
        const component = mount(
            <MemoryRouter initialEntries={['/hero/dc-superman']}>
                <Route
                    path="/hero/:id"
                    component={() => <HeroPage history={history} />} />
            </MemoryRouter>
        );

        component.find('button').prop('onClick')();
        expect(history.push).toHaveBeenCalledWith('/dc');
        expect(history.goBack).not.toHaveBeenCalled();
    });

    test(`Debe regresar a la página anterior con goBack().`, () => {

        const component = mount(
            <MemoryRouter initialEntries={['/hero/dc-superman']}>
                <Route
                    path="/hero/:id"
                    component={() => <HeroPage history={history} />} />
            </MemoryRouter>
        );

        component.find('button').prop('onClick')();
        expect(history.goBack).toHaveBeenCalledWith();
        expect(history.push).toHaveBeenCalledTimes(0);
    });

    test(`Debe llamar el Redirect si el id no existe.`, () => {

        const component = mount(
            <MemoryRouter initialEntries={['/hero/mock-hero']}>
                <Route
                    path="/hero/:id"
                    component={() => <HeroPage history={history} />} />
            </MemoryRouter>
        );

        expect(component).toMatchSnapshot();
        expect(component.text()).toBe('');
    });
});
