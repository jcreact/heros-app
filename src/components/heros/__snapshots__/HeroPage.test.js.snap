// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente 'HeroPage.jsx'. Debe llamar el Redirect si el id no existe. 1`] = `""`;

exports[`Pruebas en el componente 'HeroPage.jsx'. Debe mostrar a Superman. 1`] = `
<div
  className="MuiBox-root MuiBox-root-2"
  style={
    Object {
      "opacity": 1,
      "visibility": undefined,
    }
  }
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
    style={
      Object {
        "maxWidth": "110vh",
      }
    }
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
    >
      <div
        className="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
      >
        <img
          className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
          src="."
        />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
    >
      <div
        className="MuiBox-root MuiBox-root-3"
        style={
          Object {
            "height": "100%",
          }
        }
      >
        <h2
          className="MuiTypography-root MuiTypography-h2"
        >
          <span>
            Superman
          </span>
          <div
            className="MuiChip-root MuiBox-root MuiBox-root-4 MuiChip-colorPrimary MuiChip-outlined MuiChip-outlinedPrimary MuiChip-sizeSmall"
            onKeyDown={[Function]}
            onKeyUp={[Function]}
          >
            <span
              className="MuiChip-label MuiChip-labelSmall"
            >
              DC Comics
            </span>
          </div>
        </h2>
        <div
          className="MuiBox-root MuiBox-root-5"
        >
          <hr
            className="MuiDivider-root"
          />
        </div>
        <div
          className="MuiBox-root MuiBox-root-6"
        >
          <h5
            className="MuiTypography-root MuiTypography-h5"
          >
            <b
              style={
                Object {
                  "textTransform": "capitalize",
                }
              }
            >
              Alter ego:
               
            </b>
            Kal-El
          </h5>
        </div>
        <div
          className="MuiBox-root MuiBox-root-7"
        >
          <h5
            className="MuiTypography-root MuiTypography-h5"
          >
            <b
              style={
                Object {
                  "textTransform": "capitalize",
                }
              }
            >
              First Appearance:
               
            </b>
            Action Comics #1
          </h5>
        </div>
        <div
          className="MuiBox-root MuiBox-root-8"
        >
          <div
            className="MuiChip-root makeStyles-chips-1 MuiChip-outlined"
            onKeyDown={[Function]}
            onKeyUp={[Function]}
          >
            <div
              className="MuiAvatar-root MuiAvatar-circle MuiChip-avatar MuiAvatar-colorDefault"
            >
              KE
            </div>
            <span
              className="MuiChip-label"
            >
              Kal-El
            </span>
          </div>
        </div>
        <div
          className="MuiBox-root MuiBox-root-9"
        >
          <hr
            className="MuiDivider-root"
          />
        </div>
        <div
          className="MuiBox-root MuiBox-root-10"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <span
                className="MuiButton-startIcon MuiButton-iconSizeMedium"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11.67 3.87L9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
                  />
                </svg>
              </span>
              regresar
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
